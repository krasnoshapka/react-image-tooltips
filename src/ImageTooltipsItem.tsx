import * as React from 'react';
import {Accordion} from './Accordion';
import {imageSizeObject} from './ImageTooltips';

export interface ImageTooltipsItemProps {
  id: number;
  top: number;
  left: number;
  imageSize: imageSizeObject;
  toggle: boolean;
  children?: React.ReactNode;
  parentHandleClick: (id: number, toggle: boolean) => void
}

export const ImageTooltipsItem: React.FC<ImageTooltipsItemProps> = ({
  children,
  id,
  top,
  left,
  imageSize,
  toggle,
  parentHandleClick,
  ...props
}: ImageTooltipsItemProps) => {
  const [toggled, setToggled] = React.useState(toggle);
  if (toggled !== toggle) {
    setToggled(toggle);
  }

  const handleClick = () => {
    setToggled(!toggled);
    parentHandleClick(id, !toggled);
  };

  const cssClass = ['hotspot'];
  if (toggled) cssClass.push('toggled');

  // Validation of hotspot coordinates
  const validateCoord = (c: number, max: number) => {
    if (c <= 0) {
      console.log(`Tooltip with id ${id} has coordinate outside of image.`);
      return 1;
    }
    if (c > max) {
      console.log(`Tooltip with id ${id} has coordinate outside of image.`);
      return max;
    }
    return c;
  }
  left = validateCoord(left, imageSize.initW);
  top = validateCoord(top, imageSize.initH);

  // Calculate tooltip position based on image scaling
  const style = {
    left: `${Math.round(left * imageSize.curW / imageSize.initW)}px`,
    top: `${Math.round(top * imageSize.curH / imageSize.initH)}px`
  };

  return (
    <div className={cssClass.join(' ')} style={style} data-id={id}>
      <div className="hotspot-trigger" tabIndex={0} onClick={handleClick}>
        +
      </div>
      <Accordion toggled={toggled} {...props}>
        {children}
      </Accordion>
    </div>
  );
}
