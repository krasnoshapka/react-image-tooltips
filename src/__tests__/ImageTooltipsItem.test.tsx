import * as React from 'react';
import {_ImageTooltipsItem, _ImageTooltipsItemProps} from '../ImageTooltipsItem';
import {fireEvent, render} from "@testing-library/react";

describe('<_ImageTooltipsItem />', () => {
  const defaultProps:_ImageTooltipsItemProps = {
    top: 226,
    left: 301,
    dataId: 1,
    imageSize: {
      initW: 816,
      initH: 544,
      curW: 816,
      curH: 544
    },
    toggle: false,
    parentHandleClick: jest.fn()
  }
  const defaultChild = (
    <p>Here's some content that lies inside an tooltip.</p>
  );

  it('renders correctly', () => {
    const container = render(
      <_ImageTooltipsItem {...defaultProps}>
        {defaultChild}
      </_ImageTooltipsItem>
    );
    expect(container).toMatchSnapshot();
  });

  it('appends passed classnames', () => {
    const {container} = render(
      <_ImageTooltipsItem {...defaultProps} className={"test-trigger"}>
        {defaultChild}
      </_ImageTooltipsItem>
    );

    expect(container.firstChild!.lastChild).toHaveClass('test-trigger');
  });

  it('calls parentHandleClick', () => {
    const {container} = render(
      <_ImageTooltipsItem {...defaultProps}>
        {defaultChild}
      </_ImageTooltipsItem>
    );

    expect(defaultProps.parentHandleClick).not.toBeCalled();

    fireEvent.click(container.firstChild!.firstChild!);

    expect(defaultProps.parentHandleClick).toBeCalled();
  });

  it('validates too big coordinates', () => {
    const {container} = render(
      <_ImageTooltipsItem {...defaultProps} top={1000} left={1000}>
        {defaultChild}
      </_ImageTooltipsItem>
    );

    expect(container.firstChild).toHaveStyle({
      top: '544px',
      left: '816px'
    });
  });

  it('validates too low coordinates', () => {
    const {container} = render(
      <_ImageTooltipsItem {...defaultProps} top={-1000} left={-1000}>
        {defaultChild}
      </_ImageTooltipsItem>
    );

    expect(container.firstChild).toHaveStyle({
      top: '1px',
      left: '1px'
    });
  });

  it('scales tooltip correctly', () => {
    // Scale to 50%
    defaultProps.imageSize.curW = 408;
    defaultProps.imageSize.curH = 272;
    const {container} = render(
      <_ImageTooltipsItem {...defaultProps}>
        {defaultChild}
      </_ImageTooltipsItem>
    );

    expect(container.firstChild).toHaveStyle({
      top: '113px',
      left: '151px'
    });
  });
});
