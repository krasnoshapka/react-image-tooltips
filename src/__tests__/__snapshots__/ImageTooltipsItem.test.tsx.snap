// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<_ImageTooltipsItem /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="image-tooltips-item"
        data-id="1"
        style="left: 301px; top: 226px;"
      >
        <div
          class="image-tooltips-item__trigger "
          tabindex="0"
        >
          +
        </div>
        <div
          class="image-tooltips-item__content "
          style="height: 0px;"
        >
          <p>
            Here's some content that lies inside an tooltip.
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="image-tooltips-item"
      data-id="1"
      style="left: 301px; top: 226px;"
    >
      <div
        class="image-tooltips-item__trigger "
        tabindex="0"
      >
        +
      </div>
      <div
        class="image-tooltips-item__content "
        style="height: 0px;"
      >
        <p>
          Here's some content that lies inside an tooltip.
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
